include/lbann/utils/random.hpp

```diff
@@ -113,6 +113,86 @@ inline T fast_rand_int_pow2(Generator& g, T max) {
   return x & ((typename Generator::result_type) max);
 }
 
+// Methods for quickly generating uniformly random values in [0, 1).
+
+namespace details {
+
+// See section on converting uint64_ts to doubles in:
+// http://xoshiro.di.unimi.it/
+
+template <typename Generator>
+inline float random_float_32(Generator& g) {
+  const uint32_t r = g() >> 9;
+  return r * (1.0f / 8388608.0f);
```

Comment:
This is a cool trick. Super microoptimization (nanooptimization?): we should sample an integer from [0, 2^24). `float` has 23 mantissa bits plus an implicit leading bit.

Reply:
I don't think any of the random number generators support generating 24-bit numbers, unless we want to generate it from three 8-bit numbers, which is likely slower.

Reply:
```suggestion
  const uint32_t r = g() >> 8;
  return r * (1.0f / 16777216.0f);
```